### PR TITLE
[Pallas] Introduce GMM(torch.autograd.Function)

### PR DESCRIPTION
**Pull request by alanwaketan**
Summary:
This pull request make GMM as a torch.autograd.Function such that we can use torch.autograd.backward instead of manual backpropagation.

Test Plan:
python test/test_gmm.py

### CODE REVIEW

File: test/test_gmm.py [hunk 1]
```diff
@@ -374,6 +374,81 @@ def test_gmm_backward(self):
     # Make sure gmm doesn't fallback.
     self.assertNotIn("aten::", met.short_metrics_report())
 
+  @unittest.skipIf(xr.device_type() != 'TPU', "This test only works on TPU.")
```
Comment (JackCaoG) [inline test/test_gmm.py:377]: do you need TPU version check here?

Comment (alanwaketan) [reply test/test_gmm.py:377]: lol, good question.

Comment (alanwaketan) [reply test/test_gmm.py:377]: v2 is pretty happy on the tree.

File: test/test_gmm.py [hunk 2]
```diff
@@ -374,6 +374,81 @@ def test_gmm_backward(self):
...
+      rhs.retain_grad()
+
+      ref_out = self._reference_gmm(lhs, rhs, group_sizes)
+      ref_out.sum().backward()
+
+      torch.manual_seed(42)
+      lhs_xla = torch.rand(m, k, dtype=lhs_dtype, requires_grad=True).to("xla")
+      rhs_xla = torch.rand(
+          num_groups, k, n, dtype=rhs_dtype, requires_grad=True).to("xla")
+      lhs_xla.retain_grad()
```
Comment (JackCaoG) [inline test/test_gmm.py:401]: just curious how is this different than `lhs_xla. requires_grad  = True`?

File: test/test_gmm.py [hunk 3]
```diff
@@ -374,6 +374,81 @@ def test_gmm_backward(self):
...
+      out.sum().backward()
+
+      self.assertTrue(torch.allclose(ref_out, out.cpu()))
+      self.assertTrue(torch.allclose(lhs.grad, lhs_xla.grad.cpu()))
+      self.assertTrue(torch.allclose(rhs.grad, rhs_xla.grad.cpu()))
+
+    # Make sure gmm doesn't fallback.
+    self.assertNotIn("aten::", met.short_metrics_report())
+
+  @unittest.skipIf(xr.device_type() != 'TPU', "This test only works on TPU.")
```
Comment (JackCaoG) [inline test/test_gmm.py:414]: ditto, otherwise I think it will break v2 CI

**Review by JackCaoG — APPROVED** [review]

File: test/test_gmm.py (hunk 2 above)
Comment (alanwaketan) [reply test/test_gmm.py:401]: It make the .grad appears. Otherwise, it's None.

File: test/test_gmm.py (hunk 1 above)
Comment (JackCaoG) [reply test/test_gmm.py:377]: interesting.. I thought pallas is not supported on v2.

File: test/test_gmm.py (hunk 3 above)
Comment (alanwaketan) [reply test/test_gmm.py:414]: Actually, the CI somehow is happy. Maybe it just don't like the others.

File: test/test_gmm.py (hunk 1 above)
Comment (alanwaketan) [reply test/test_gmm.py:377]: gmm is just mm... The kernel is simple... Other fuses softmax, etc...

**Comment by alanwaketan** [comment]
Thanks Jack for approving.

**Comment by alanwaketan** [comment]
Skip GPU tests to move fast.
